actors/builtin/paych/paych_test.go

```diff
@@ -23,156 +25,705 @@ func TestPaymentChannelActor_Constructor(t *testing.T) {
+	var st1 State
+
+	rt.GetState(&st1)
+	rt.SetCaller(st1.From, builtin.AccountActorCodeID)
+
+	var st2 State
+	mergeTo := st1.LaneStates[0]
+	mergeFrom := st1.LaneStates[1]
+
+	// Note sv.Amount = 4
+	sv.Lane = mergeTo.ID
+	mergeNonce := mergeTo.Nonce + 10
+
+	merges := []Merge{{Lane: mergeFrom.ID, Nonce: mergeNonce}}
+	sv.Merges = merges
+
+	ucp := &UpdateChannelStateParams{Sv: *sv}
+	rt.ExpectValidateCallerAddr(st1.From, st1.To)
+	_ = rt.Call(actor.UpdateChannelState, ucp).(*adt.EmptyValue)
+	rt.Verify()
+
+	rt.GetState(&st2)
+	newMergeTo := st2.LaneStates[0]
+	newMergeFrom := st2.LaneStates[1]
+	require.NotNil(t, newMergeTo)
+	require.NotNil(t, newMergeFrom)
+
+	assert.Equal(t, int(mergeNonce), int(newMergeFrom.Nonce))
+	assert.Equal(t, mergeFrom.Redeemed, newMergeFrom.Redeemed)
+	assert.Equal(t, int(sv.Nonce), int(newMergeTo.Nonce))
+	assert.Equal(t, sv.Amount, newMergeTo.Redeemed)
+
+	redeemed := big.Add(mergeFrom.Redeemed, mergeTo.Redeemed)
+	expDelta := big.Sub(sv.Amount, redeemed)
+	expSendAmt := big.Add(st1.ToSend, expDelta)
+	assert.Equal(t, expSendAmt, st2.ToSend)
+	assert.Len(t, st2.LaneStates, numLanes)
+}
+
+func TestActor_UpdateChannelStateMergeFailure(t *testing.T) {
+	testCases := []struct {
+		name                           string
+		balance                        int64
+		lane, voucherNonce, mergeNonce uint64
+		expExitCode                    exitcode.ExitCode
+	}{
+		{
+			name: "fails: merged lane in voucher has outdated nonce, cannot redeem",
+			lane: 1, voucherNonce: 10, mergeNonce: 1,
+			expExitCode: exitcode.ErrIllegalArgument,
+		},
+		{
+			name: "fails: voucher has an outdated nonce, cannot redeem",
+			lane: 1, voucherNonce: 0, mergeNonce: 10,
+			expExitCode: exitcode.ErrIllegalArgument,
+		},
+		{
+			name: "fails: not enough funds in channel to cover voucher",
+			lane: 1, balance: 1, voucherNonce: 10, mergeNonce: 10,
+			expExitCode: exitcode.ErrIllegalState,
+		},
+		{
+			name: "fails: voucher cannot merge lanes into its own lane",
+			lane: 0, voucherNonce: 10, mergeNonce: 10,
+			expExitCode: exitcode.ErrIllegalArgument,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			rt, actor, sv := requireCreateChannelWithLanes(t, context.Background(), 2)
+			if tc.balance > 0 {
+				rt.SetBalance(abi.NewTokenAmount(tc.balance))
+			}
+
+			var st1 State
+			rt.GetState(&st1)
+			mergeTo := st1.LaneStates[0]
+			mergeFrom := st1.LaneStates[tc.lane]
+
+			sv.Lane = mergeTo.ID
+			sv.Nonce = tc.voucherNonce
+			merges := []Merge{{Lane: mergeFrom.ID, Nonce: tc.mergeNonce}}
+			sv.Merges = merges
+			ucp := &UpdateChannelStateParams{Sv: *sv}
+
+			rt.SetCaller(st1.From, builtin.AccountActorCodeID)
+			rt.ExpectValidateCallerAddr(st1.From, st1.To)
+			rt.ExpectAbort(tc.expExitCode, func() {
+				rt.Call(actor.UpdateChannelState, ucp)
+			})
+
+		})
+	}
+	t.Run("When lane doesn't exist, fails with: voucher specifies invalid merge lane 999", func(t *testing.T) {
```

Comment:
I've changed some of the descriptions to reflect the expected reason for failure.  This helped me identify some false positives.